server/src/main/java/org/opensearch/cluster/decommission/DecommissionService.java

```diff
@@ -0,0 +1,321 @@
+import org.opensearch.cluster.coordination.NodeRemovalClusterStateTaskExecutor;
+import org.opensearch.cluster.metadata.DecommissionAttributeMetadata;
+import org.opensearch.cluster.metadata.Metadata;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.routing.allocation.AllocationService;
+import org.opensearch.cluster.routing.allocation.decider.AwarenessAllocationDecider;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.Priority;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.common.io.stream.StreamInput;
+import org.opensearch.common.settings.ClusterSettings;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.threadpool.ThreadPool;
+import org.opensearch.transport.TransportException;
+import org.opensearch.transport.TransportResponse;
+import org.opensearch.transport.TransportResponseHandler;
+import org.opensearch.transport.TransportService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+public class DecommissionService implements ClusterStateApplier {
+
+    private static final Logger logger = LogManager.getLogger(DecommissionService.class);
+
+    private final ClusterService clusterService;
+    private final TransportService transportService;
+    private final NodeRemovalClusterStateTaskExecutor nodeRemovalExecutor;
+    private final ThreadPool threadPool;
+    private final DecommissionHelper decommissionHelper;
+    private ClusterState clusterState;
+    private volatile List<String> awarenessAttributes;
+
+    @Inject
+    public DecommissionService(
+        Settings settings,
+        ClusterSettings clusterSettings,
+        ClusterService clusterService,
+        TransportService transportService,
+        ThreadPool threadPool,
+        AllocationService allocationService
+    ) {
+        this.clusterService = clusterService;
+        this.transportService = transportService;
+        this.threadPool = threadPool;
+        this.nodeRemovalExecutor = new NodeRemovalClusterStateTaskExecutor(allocationService, logger);
+        this.decommissionHelper = new DecommissionHelper(
+            clusterService,
+            nodeRemovalExecutor
+        );
+        this.awarenessAttributes = AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING.get(settings);
+        clusterSettings.addSettingsUpdateConsumer(
+            AwarenessAllocationDecider.CLUSTER_ROUTING_ALLOCATION_AWARENESS_ATTRIBUTE_SETTING,
+            this::setAwarenessAttributes
+        );
+    }
+
+    List<String> getAwarenessAttributes() {
+        return awarenessAttributes;
+    }
+
+    private void setAwarenessAttributes(List<String> awarenessAttributes) {
+        this.awarenessAttributes = awarenessAttributes;
+    }
+
+    public void initiateAttributeDecommissioning(
```

Comment:
What would the onResponse and onFailure look like for the below listener?

Reply:
As soon as the `decommissionedAttribute` is registered with status as `DECOMMISSION_INIT` we acknowledge the response. In case of errors while doing that due to validators, etc, false acknowledgement is returned

Reply:
nit : we don't need to pass `state` here .

Reply:
Should this be `synchronized` to ensure we don't have concurrent decommission operations running?